flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/runtime/stream/table/ValuesITCase.java

```diff
@@ -0,0 +1,309 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.runtime.stream.table;
+
+import org.apache.flink.table.api.DataTypes;
+import org.apache.flink.table.api.Table;
+import org.apache.flink.table.planner.factories.utils.TestCollectionTableFactory;
+import org.apache.flink.table.planner.runtime.utils.StreamingTestBase;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.types.Row;
+
+import org.junit.Test;
+
+import java.math.BigDecimal;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.junit.Assert.assertThat;
+
+/**
+ * End to end tests for {@link org.apache.flink.table.api.TableEnvironment#fromValues}.
+ */
+public class ValuesITCase extends StreamingTestBase {
+	@Test
+	public void testTypeConvesrsions() throws Exception {
+		List<Row> data = Arrays.asList(
+			Row.of(
+				1,
+				"ABC",
+				java.sql.Timestamp.valueOf("2000-12-12 12:30:57.12"),
+				Row.of(1, new byte[]{1, 2}, "ABC", Arrays.asList(1, 2, 3))),
+			Row.of(
+				Math.PI,
+				"ABC",
+				LocalDateTime.parse("2000-12-12T12:30:57.123456"),
+				Row.of(Math.PI, new byte[]{2, 3}, "ABC", Arrays.asList(1L, 2L, 3L))),
+			Row.of(
+				3.1f,
+				"DEF",
+				LocalDateTime.parse("2000-12-12T12:30:57.1234567"),
+				Row.of(3.1f, new byte[]{3}, "DEF", Arrays.asList(1D, 2D, 3D))),
+			Row.of(
+				99L,
+				"DEFG",
+				LocalDateTime.parse("2000-12-12T12:30:57.12345678"),
+				Row.of(99L, new byte[]{3, 4}, "DEFG", Arrays.asList(1f, 2f, 3f))),
+			Row.of(
+				0d,
+				"D",
+				LocalDateTime.parse("2000-12-12T12:30:57.123"),
+				Row.of(0d, new byte[]{4}, "D", Arrays.asList(1, 2, 3)))
+		);
+
+		DataType rowType = DataTypes.ROW(
+			DataTypes.FIELD("a", DataTypes.DECIMAL(10, 2).notNull()),
+			DataTypes.FIELD("b", DataTypes.CHAR(4).notNull()),
+			DataTypes.FIELD("c", DataTypes.TIMESTAMP(5).notNull()),
+			DataTypes.FIELD(
+				"row",
+				DataTypes.ROW(
+					DataTypes.FIELD("a", DataTypes.DECIMAL(10, 3)),
+					DataTypes.FIELD("b", DataTypes.BINARY(2)),
+					DataTypes.FIELD("c", DataTypes.CHAR(5).notNull()),
+					DataTypes.FIELD("d", DataTypes.ARRAY(DataTypes.DECIMAL(10, 2))))
+			)
+		);
+
+		Table t = tEnv().fromValues(
+			rowType,
+			data
+		);
+
+		TestCollectionTableFactory.reset();
+		tEnv().sqlUpdate(
+			"CREATE TABLE SinkTable(" +
+				"a DECIMAL(10, 2) NOT NULL, " +
+				"b CHAR(4) NOT NULL," +
+				"c TIMESTAMP(4) NOT NULL," +
```

Comment:
There is a mismatch of values and sink. `TIMESTAMP(4)` vs `DataTypes.TIMESTAMP(5)`